python/tskit/metadata.py

```diff
@@ -120,6 +123,191 @@ def decode(self, data: bytes) -> bytes:
         return data
 
 
+def binary_format_validator(validator, types, instance, schema):
+    # Make sure the normal type validation gets done
+    yield from jsonschema._validators.type(validator, types, instance, schema)
+
+    # Non-composite types must have a binaryFormat
+    if (
+        validator.is_type(instance, "object")
+        and (instance.get("type") not in (None, "object", "array", "null"))
+        and "binaryFormat" not in instance
+    ):
+        yield jsonschema.ValidationError(
+            f"{instance['type']} type must have binaryFormat set"
+        )
+
+
+StructCodecSchemaValidator = jsonschema.validators.extend(
+    TSKITMetadataSchemaValidator, {"type": binary_format_validator}
+)
+META_SCHEMA: Mapping[str, Any] = copy.deepcopy(StructCodecSchemaValidator.META_SCHEMA)
+# No union types
+META_SCHEMA["properties"]["type"] = {"$ref": "#/definitions/simpleTypes"}
+META_SCHEMA["definitions"]["root"]["properties"]["type"] = {
+    "$ref": "#/definitions/simpleTypes"
+}
+# No hetrogeneous arrays
+META_SCHEMA["properties"]["items"] = {"$ref": "#/definitions/root"}
+META_SCHEMA["definitions"]["root"]["properties"]["items"] = {
+    "$ref": "#/definitions/root"
+}
+# binaryFormat matches regex
+META_SCHEMA["properties"]["binaryFormat"] = {
+    "type": "string",
+    "pattern": r"^([cbB\?hHiIlLqQfd]|\d*[sp])$",
+}
+META_SCHEMA["definitions"]["root"]["properties"]["binaryFormat"] = {
+    "type": "string",
+    "pattern": r"^[cbB\?hHiIlLqQfd]|\d*[sp]$",
+}
+StructCodecSchemaValidator.META_SCHEMA = META_SCHEMA
+
+
+class StructCodec(AbstractMetadataCodec):
+    """
+    Codec that encodes data using struct. Note that this codec has extra restrictions
+    Namely that object keys must be fixed (all present and no extra); each entry should
+    have a binaryFormat; that arrays are homogeneous and that types are not unions.
+    """
+
+    @classmethod
+    def order_by_index(cls, obj, do_sort=False):
+        """
+        Take a schema and recursively convert any dict that is under the key
+        name ``properties`` to an OrderedDict.
+        """
+        if isinstance(obj, collections.abc.Mapping):
+            keys = obj.keys()
+            if do_sort:
+                keys = sorted(keys)
+            items = {k: cls.order_by_index(obj[k], k == "properties") for k in keys}
+            if do_sort:
+                return collections.OrderedDict(items)
+            else:
+                return dict(items)
+        elif isinstance(obj, list) or isinstance(obj, tuple):
+            return [cls.order_by_index(v, False) for v in obj]
+        else:
+            return obj
+
+    def __init__(self, schema: Mapping[str, Any]) -> None:
+        # This codec requires that all properties are required and additional ones
+        # not allowed. Rather than get schema authors to repeat that everywhere
+        # we add it here
+        def enforce_fixed_properties(obj):
```

Comment:
Be helpful to note here why we can't do this with the metaschema.

Reply:
Note added in https://github.com/tskit-dev/tskit/pull/601/commits/09cbea7cde5ced87bb5229f921f39cf32ca56de9